File: android/src/main/java/expo/modules/entrig/EntrigModule.kt
package expo.modules.entrig

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.util.Log
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import com.entrig.sdk.Entrig
import com.entrig.sdk.models.EntrigConfig
import com.entrig.sdk.models.NotificationEvent

class EntrigModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw IllegalStateException("React context is null")

  private val currentActivity: Activity?
    get() = appContext.currentActivity

  override fun definition() = ModuleDefinition {
    Name("Entrig")

    // Defines event names that the module can send to JavaScript.
    Events("onForegroundNotification", "onNotificationOpened")

    OnCreate {
      // Set up SDK listeners when module is created
      Entrig.setOnForegroundNotificationListener { notification ->
        Log.d("EntrigModule", "Foreground notification: ${notification.toMap()}")
        sendEvent("onForegroundNotification", notification.toMap())
      }

      Entrig.setOnNotificationOpenedListener { notification ->
        Log.d("EntrigModule", "Notification opened: ${notification.toMap()}")
        sendEvent("onNotificationOpened", notification.toMap())
      }
    }

    // Initialize SDK
    AsyncFunction("init") { config: Map<String, Any?>, promise: Promise ->
      val apiKey = config["apiKey"] as? String
      if (apiKey.isNullOrEmpty()) {
        promise.reject("INVALID_API_KEY", "API key is required and cannot be empty", null)
        return@AsyncFunction
      }

      val handlePermission = config["handlePermission"] as? Boolean ?: true
      val entrigConfig = EntrigConfig(
        apiKey = apiKey,
        handlePermission = handlePermission
      )

      // Pass application context to ensure lifecycle callbacks are registered
      val appContext = context.applicationContext
      Entrig.initialize(appContext, entrigConfig) { success, error ->
        if (success) {
          promise.resolve(null)
        } else {
          promise.reject("INIT_ERROR", error ?: "Failed to initialize SDK", null)
        }
      }
    }

    // Register user
    AsyncFunction("register") { userId: String, promise: Promise ->
      val activity = currentActivity
      if (activity == null) {
        promise.reject("NO_ACTIVITY", "Activity not available", null)
        return@AsyncFunction
      }

      Entrig.register(userId, activity) { success, error ->
        if (success) {
          promise.resolve(null)
        } else {
          promise.reject("REGISTER_ERROR", error ?: "Registration failed", null)
        }
      }
    }

    // Request permission
    AsyncFunction("requestPermission") { promise: Promise ->
      val activity = currentActivity
      if (activity == null) {
        promise.reject("NO_ACTIVITY", "Activity not available", null)
        return@AsyncFunction
      }

      Entrig.requestPermission(activity) { granted ->
        promise.resolve(granted)
      }
    }

    // Unregister
    AsyncFunction("unregister") { promise: Promise ->
      Entrig.unregister { success, error ->
        if (success) {
          promise.resolve(null)
        } else {
          promise.reject("UNREGISTER_ERROR", error ?: "Unregistration failed", null)
        }
      }
    }

    // Get initial notification
    AsyncFunction("getInitialNotification") { promise: Promise ->
      val initialNotification = Entrig.getInitialNotification()
      promise.resolve(initialNotification?.toMap())
    }
  }
}

File: ios/EntrigModule.swift
import ExpoModulesCore
import UserNotifications
import EntrigSDK

public class EntrigModule: Module {
  public func definition() -> ModuleDefinition {
    Name("Entrig")

    // Defines event names that the module can send to JavaScript.
    Events("onForegroundNotification", "onNotificationOpened")

    // Initialize SDK
    AsyncFunction("init") { (config: [String: Any]) -> Void in
      guard let apiKey = config["apiKey"] as? String, !apiKey.isEmpty else {
        throw Exception(name: "INVALID_API_KEY", description: "API key is required and cannot be empty")
      }

      let handlePermission = config["handlePermission"] as? Bool ?? true
      let entrigConfig = EntrigConfig(apiKey: apiKey, handlePermission: handlePermission)

      // Set up SDK listeners
      Entrig.setOnForegroundNotificationListener(self)
      Entrig.setOnNotificationOpenedListener(self)

      return try await withCheckedThrowingContinuation { continuation in
        Entrig.configure(config: entrigConfig) { success, error in
          if success {
            continuation.resume()
          } else {
            continuation.resume(throwing: Exception(name: "INIT_ERROR", description: error ?? "Failed to initialize SDK"))
          }
        }
      }
    }

    // Register user
    AsyncFunction("register") { (userId: String) -> Void in
      return try await withCheckedThrowingContinuation { continuation in
        Entrig.register(userId: userId, sdk: "expo") { success, error in
          if success {
            continuation.resume()
          } else {
            continuation.resume(throwing: Exception(name: "REGISTER_ERROR", description: error ?? "Registration failed"))
          }
        }
      }
    }

    // Request permission
    AsyncFunction("requestPermission") { () -> Bool in
      return try await withCheckedThrowingContinuation { continuation in
        Entrig.requestPermission { granted, error in
          if let error = error {
            continuation.resume(throwing: Exception(name: "PERMISSION_ERROR", description: error.localizedDescription))
          } else {
            continuation.resume(returning: granted)
          }
        }
      }
    }

    // Unregister
    AsyncFunction("unregister") { () -> Void in
      return try await withCheckedThrowingContinuation { continuation in
        Entrig.unregister { success, error in
          if success {
            continuation.resume()
          } else {
            continuation.resume(throwing: Exception(name: "UNREGISTER_ERROR", description: error ?? "Unregistration failed"))
          }
        }
      }
    }

    // Get initial notification
    AsyncFunction("getInitialNotification") { () -> [String: Any]? in
      if let event = Entrig.getInitialNotification() {
        return [
          "title": event.title ?? "",
          "body": event.body ?? "",
          "data": event.data ?? [:],
          "isForeground": false
        ]
      }
      return nil
    }
  }

  // Helper method to send notification events to JavaScript
  private func sendNotificationToJS(event: NotificationEvent, isForeground: Bool) {
    let payload: [String: Any] = [
      "title": event.title ?? "",
      "body": event.body ?? "",
      "data": event.data ?? [:],
      "isForeground": isForeground
    ]

    if isForeground {
      sendEvent("onForegroundNotification", payload)
    } else {
      sendEvent("onNotificationOpened", payload)
    }
  }
}

// MARK: - SDK Listeners
extension EntrigModule: OnNotificationReceivedListener {
  public func onNotificationReceived(_ event: NotificationEvent) {
    sendNotificationToJS(event: event, isForeground: true)
  }
}

extension EntrigModule: OnNotificationClickListener {
  public func onNotificationClick(_ event: NotificationEvent) {
    sendNotificationToJS(event: event, isForeground: false)
  }
}
